File: tests/src/com.google.protobuf/protobuf-javalite/3.25.3/src/test/java/com_google_protobuf/protobuf_javalite/UnsafeUtilInnerAndroid32MemoryAccessorTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_protobuf.protobuf_javalite;

import static java.lang.invoke.MethodType.methodType;
import static org.assertj.core.api.Assertions.assertThat;

import com.google.protobuf.ByteString;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.reflect.Field;
import org.junit.jupiter.api.Test;

public class UnsafeUtilInnerAndroid32MemoryAccessorTest {
    private static final String ANDROID_32_MEMORY_ACCESSOR_CLASS_NAME =
            "com.google.protobuf.UnsafeUtil$Android32MemoryAccessor";
    public static final Object STATIC_VALUE = new Object();

    @Test
    void android32MemoryAccessorReadsStaticFieldValue() throws Throwable {
        ClassLoader protobufClassLoader = ByteString.class.getClassLoader();
        Class<?> android32MemoryAccessorClass = Class.forName(
                ANDROID_32_MEMORY_ACCESSOR_CLASS_NAME,
                true,
                protobufClassLoader
        );
        MethodHandles.Lookup android32MemoryAccessorLookup = MethodHandles.privateLookupIn(
                android32MemoryAccessorClass,
                MethodHandles.lookup()
        );
        MethodHandle constructor = android32MemoryAccessorLookup.findConstructor(
                android32MemoryAccessorClass,
                methodType(void.class, sun.misc.Unsafe.class)
        );
        Object memoryAccessor = constructor.invoke((sun.misc.Unsafe) null);
        MethodHandle getStaticObject = android32MemoryAccessorLookup.findVirtual(
                android32MemoryAccessorClass,
                "getStaticObject",
                methodType(Object.class, Field.class)
        );

        Object value = getStaticObject.invoke(
                memoryAccessor,
                UnsafeUtilInnerAndroid32MemoryAccessorTest.class.getDeclaredField("STATIC_VALUE")
        );

        assertThat(value).isSameAs(STATIC_VALUE);
    }
}

File: tests/src/org.apache.pekko/pekko-protobuf-v3_2.13/1.2.0/src/test/scala/org_apache_pekko/pekko_protobuf_v3_2_13/UnsafeUtilInnerAndroid32MemoryAccessorTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_pekko.pekko_protobuf_v3_2_13

import java.lang.Void.TYPE
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType.methodType

import org.apache.pekko.protobufv3.internal.ByteString
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class UnsafeUtilInnerAndroid32MemoryAccessorTest {
  @Test
  def getStaticObjectReadsStaticFieldValue(): Unit = {
    val accessorClass: Class[_] = Class.forName(
      "org.apache.pekko.protobufv3.internal.UnsafeUtil$Android32MemoryAccessor",
      true,
      classOf[ByteString].getClassLoader
    )
    val lookup: MethodHandles.Lookup = MethodHandles.privateLookupIn(
      accessorClass,
      MethodHandles.lookup()
    )
    val constructor: MethodHandle = lookup.findConstructor(
      accessorClass,
      methodType(TYPE, classOf[sun.misc.Unsafe])
    )
    val accessor: Object = constructor
      .invokeWithArguments(null.asInstanceOf[Object])
      .asInstanceOf[Object]
    val getStaticObject: MethodHandle = lookup.findVirtual(
      accessorClass,
      "getStaticObject",
      methodType(classOf[Object], classOf[java.lang.reflect.Field])
    )

    val emptyField: java.lang.reflect.Field = classOf[ByteString].getField("EMPTY")
    val value: Object = getStaticObject
      .invokeWithArguments(accessor, emptyField)
      .asInstanceOf[Object]

    assertThat(value).isSameAs(ByteString.EMPTY)
  }
}
